### ISSUE TITLE
Screen shows Snowflakes but the first few lines  (BSP-324)

### ISSUE THREAD

**Issue by hayschan** [espressif/esp-bsp#173]
## TL;DR

TL;DR: After updating to the latest Waveshare 7-inch screen component, the UI started displaying snowflakes and flickering. The error log indicates a task watchdog timeout, with the LVGL task failing to reset the watchdog in time.

## Version

Waveshare 7inch screen component: `v1.0.1`

### Before

When I was using Waveshare 7inch screen component `v1.0.0`, the UI works normally as expected. 

### After

Right after the change to `v1.0.1`, there are **snowflakes** on the screen. At the very top, there is a few lines of pixel that display normally. But other parts of the screen remains white and flickering.

![image](https://github.com/espressif/esp-bsp/assets/25737801/5d18e864-ce27-4657-b7ec-74d8e813200f)

The UI does not respond to touch screen either.

## Code

`main.c`
```c
/*
 * SPDX-FileCopyrightText: 2022 Espressif Systems (Shanghai) CO LTD
 *
 * SPDX-License-Identifier: CC0-1.0
 */

#include "esp_log.h"
#include "bsp/ws_7inch.h"
#include "lvgl.h"
#include "ui/ui.h"


#define TAG "ESP-EXAMPLE"

/*******************************************************************************
* Private functions
*******************************************************************************/

// *INDENT-OFF*
void app_lvgl_display(void)
{
    bsp_display_lock(0);

    ui_init();

    bsp_display_unlock();
}

void app_main(void)
{
    /* Initialize I2C (for touch) */
    bsp_i2c_init();

    /* Initialize display and LVGL */
    bsp_display_start();

    /* Turn on display backlight */
    bsp_display_backlight_on();

    /* Add and show objects on display */
    app_lvgl_display();

    ESP_LOGI(TAG, "Example initialization done.");
}
// *INDENT-ON*
```

## Error log

I have also check the serial monitor. It seems that LVGL task is triggering watchdog reset. Then, it resets itself and keeps repeating this process again and again. 

Check out `I (1787)`, ESP32 prints out `ESP-EXAMPLE: Example initialization done.` which means that the initialization in `app_main()` was executed normally.

```python
I (1447) LVGL: Starting LVGL task
I (1447) gpio: GPIO[9]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (1477) PWM_var: PWM_var = 0.00%
I (1487) gpio: GPIO[11]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (1517) gpio: GPIO[10]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:2 
I (1517) GT911: TouchPad_ID:0x39,0x31,0x31
I (1517) GT911: TouchPad_Config_Version:91
I (1787) gpio: GPIO[14]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 
I (1787) ESP-EXAMPLE: Example initialization done.
E (6957) task_wdt: Task watchdog got triggered. The following tasks/users did not reset the watchdog in time:
E (6957) task_wdt:  - IDLE (CPU 0)
E (6957) task_wdt: Tasks currently running:
E (6957) task_wdt: CPU 0: LVGL task
E (6957) task_wdt: CPU 1: IDLE
E (6957) task_wdt: Print CPU 0 (current core) backtrace


Backtrace: 0x42045D0E:0x3FC94600 0x42045EC4:0x3FC94620 0x40377139:0x3FC94640 0x4201BABC:0x3FCF7BA0 0x4201C0C5:0x3FCF7BC0 0x4201C21B:0x3FCF7C30 0x4201C328:0x3FCF7C60 0x4201C4D0:0x3FCF7C80 0x4202C041:0x3FCF7CB0 0x4202C0EF:0x3FCF7CD0 0x42010FDC:0x3FCF7CF0 0x4037E955:0x3FCF7D10
0x42045d0e: task_wdt_timeout_handling at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/esp_system/task_wdt.c:486 (discriminator 3)

0x42045ec4: task_wdt_isr at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/esp_system/task_wdt.c:586

0x40377139: _xt_lowint1 at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/freertos/FreeRTOS-Kernel/portable/xtensa/xtensa_vectors.S:1117

0x4201babc: draw_buf_flush at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:1198
0x4201c0c5: refr_area_part at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:712

0x4201c21b: refr_area at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:594 (discriminator 2)

0x4201c328: refr_invalid_areas at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:532

0x4201c4d0: _lv_disp_refr_timer at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:324

0x4202c041: lv_timer_exec at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/misc/lv_timer.c:313 
(discriminator 2)

0x4202c0ef: lv_timer_handler at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/misc/lv_timer.c:109

0x42010fdc: lvgl_port_task at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/espressif__esp_lvgl_port/esp_lvgl_port.c:441

0x4037e955: vPortTaskWrapper at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/freertos/FreeRTOS-Kernel/portable/xtensa/port.c:151


E (11957) task_wdt: Task watchdog got triggered. The following tasks/users did not reset the watchdog in time:
E (11957) task_wdt:  - IDLE (CPU 0)
E (11957) task_wdt: Tasks currently running:
E (11957) task_wdt: CPU 0: LVGL task
E (11957) task_wdt: CPU 1: IDLE
E (11957) task_wdt: Print CPU 0 (current core) backtrace


Backtrace: 0x42045D0E:0x3FC94600 0x42045EC4:0x3FC94620 0x40377139:0x3FC94640 0x4201BAB9:0x3FCF7BA0 0x4201C0C5:0x3FCF7BC0 0x4201C21B:0x3FCF7C30 0x4201C328:0x3FCF7C60 0x4201C4D0:0x3FCF7C80 0x4202C041:0x3FCF7CB0 0x4202C0EF:0x3FCF7CD0 0x42010FDC:0x3FCF7CF0 0x4037E955:0x3FCF7D10
0x42045d0e: task_wdt_timeout_handling at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/esp_system/task_wdt.c:486 (discriminator 3)

0x42045ec4: task_wdt_isr at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/esp_system/task_wdt.c:586

0x40377139: _xt_lowint1 at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/freertos/FreeRTOS-Kernel/portable/xtensa/xtensa_vectors.S:1117

0x4201bab9: draw_buf_flush at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:1199 (discriminator 1)

0x4201c0c5: refr_area_part at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:712

0x4201c21b: refr_area at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:594 (discriminator 2)

0x4201c328: refr_invalid_areas at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:532

0x4201c4d0: _lv_disp_refr_timer at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:324

0x4202c041: lv_timer_exec at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/misc/lv_timer.c:313 
(discriminator 2)

0x4202c0ef: lv_timer_handler at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/misc/lv_timer.c:109

0x42010fdc: lvgl_port_task at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/espressif__esp_lvgl_port/esp_lvgl_port.c:441

0x4037e955: vPortTaskWrapper at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/freertos/FreeRTOS-Kernel/portable/xtensa/port.c:151

```

**Comment by hayschan** [comment]
## Summary

Based on the provided log, the LVGL task watchdog reset is being triggered. Here's the relevant information from the log:

```python
E (6612) task_wdt: Task watchdog got triggered. The following tasks/users did not reset the watchdog in time:
E (6612) task_wdt:  - IDLE (CPU 0)
E (6612) task_wdt: Tasks currently running:
E (6612) task_wdt: CPU 0: LVGL task
E (6612) task_wdt: CPU 1: IDLE
```

According to the log, the LVGL task on CPU 0 is not resetting the watchdog in time. The watchdog is a mechanism that ensures tasks are executing within the expected timeframe. When a task fails to reset the watchdog, it indicates that the task is taking too long to complete, which can be an indication of a potential issue in the code.

The backtrace information provides additional details about the code execution at the time the watchdog reset occurred:

```python
Backtrace: 0x42043A8A:0x3FC93EF0 0x42043C40:0x3FC93F10 0x40377101:0x3FC93F30 0x4201AA86:0x3FCEADD0 0x4201B0BD:0x3FCEADF0 0x4201B213:0x3FCEAE60 0x4201B320:0x3FCEAE90 0x4201B4EC:0x3FCEAEB0 0x4202B961:0x3FCEAEE0 0x4202BA1F:0x3FCEAF00 0x42010184:0x3FCEAF20 0x4037E29D:0x3FCEAF40
```

The backtrace shows the sequence of function calls leading up to the watchdog reset. Here are some of the relevant functions in the backtrace related to LVGL:

- `draw_buf_flush` at `E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:1198`
- `refr_area_part` at `E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:712`
- `refr_area` at `E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:594`
- `refr_invalid_areas` at `E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:532`
- `_lv_disp_refr_timer` at `E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:324`
- `lv_timer_exec` at `E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/misc/lv_timer.c:313`
- `lv_timer_handler` at `E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/misc/lv_timer.c:109`
- `lvgl_port_task` at `E:/esp_fryer_highend/ESP-fryer-highend/managed_components/espressif__esp_lvgl_port/esp_lvgl_port.c:441`

These functions are part of the LVGL library and are involved in the rendering and refreshing of the LVGL UI. The issue may be related to the LVGL task being overloaded or not completing its tasks within the allocated time.

To identify the specific part of the code causing the LVGL task watchdog reset, you can review the relevant LVGL functions mentioned in the backtrace. Look for any potential bottlenecks or long-running operations within those functions that could be causing the task to take too long.

Additionally, enable LVGL debug logging to get more detailed information about the execution flow and any potential errors or warnings that may help pinpoint the problem area.

Keep in mind that troubleshooting complex issues like this often requires a deep understanding of the codebase and the specific hardware setup. If you're unable to identify the cause of the issue, consider reaching out to the LVGL community or the developers of the ESP-fryer-highend project for further assistance.

## Error Log

```python
I (0) cpu_start: App cpu up.
I (996) cpu_start: Pro cpu start user code
I (997) cpu_start: cpu freq: 160000000 Hz
I (997) cpu_start: Application information:
I (1000) cpu_start: Project name:     ESP-fryer-highend
I (1005) cpu_start: App version:      v1.2.0-12-g939ccbb-dirty
I (1012) cpu_start: Compile time:     May 19 2023 16:18:03
I (1018) cpu_start: ELF file SHA256:  e6573f853a2b17f6...
I (1024) cpu_start: ESP-IDF:          v5.0-dirty
I (1030) heap_init: Initializing. RAM available for dynamic allocation:
I (1037) heap_init: At 3FC97AA0 len 00051C70 (327 KiB): D/IRAM
I (1043) heap_init: At 3FCE9710 len 00005724 (21 KiB): STACK/DRAM
I (1050) heap_init: At 3FCF0000 len 00008000 (32 KiB): DRAM
I (1056) heap_init: At 600FE010 len 00001FF0 (7 KiB): RTCRAM
I (1064) spi_flash: detected chip: generic
I (1068) spi_flash: flash io: dio
W (1072) pcnt(legacy): legacy driver is deprecated, please migrate to `driver/pulse_cnt.h`
W (1081) ADC: legacy driver is deprecated, please migrate to `esp_adc/adc_oneshot.h`
I (1089) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.
I (10) LVGL: Starting LVGL task
I (1112) gpio: GPIO[9]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (1122) gpio: GPIO[11]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (1152) gpio: GPIO[10]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:2 
I (1152) GT911: TouchPad_ID:0x39,0x31,0x31
I (1152) GT911: TouchPad_Config_Version:91
I (1422) ESP-EXAMPLE: Example initialization done.
E (6612) task_wdt: Task watchdog got triggered. The following tasks/users did not reset the watchdog in time:
E (6612) task_wdt:  - IDLE (CPU 0)
E (6612) task_wdt: Tasks currently running:
E (6612) task_wdt: CPU 0: LVGL task
E (6612) task_wdt: CPU 1: IDLE
E (6612) task_wdt: Print CPU 0 (current core) backtrace


Backtrace: 0x42043A8A:0x3FC93EF0 0x42043C40:0x3FC93F10 0x40377101:0x3FC93F30 0x4201AA86:0x3FCEADD0 0x4201B0BD:0x3FCEADF0 0x4201B213:0x3FCEAE60 0x4201B320:0x3FCEAE90 0x4201B4EC:0x3FCEAEB0 0x4202B961:0x3FCEAEE0 0x4202BA1F:0x3FCEAF00 0x42010184:0x3FCEAF20 0x4037E29D:0x3FCEAF40
0x42043a8a: task_wdt_timeout_handling at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/esp_system/task_wdt.c:486 (discriminator 3)

0x42043c40: task_wdt_isr at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/esp_system/task_wdt.c:586

0x40377101: _xt_lowint1 at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/freertos/FreeRTOS-Kernel/portable/xtensa/xtensa_vectors.S:1117

0x4201aa86: draw_buf_flush at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:1198
0x4201b0bd: refr_area_part at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:712

0x4201b213: refr_area at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:594 (discriminator 2)

0x4201b320: refr_invalid_areas at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:532

0x4201b4ec: _lv_disp_refr_timer at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/core/lv_refr.c:324

0x4202b961: lv_timer_exec at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/misc/lv_timer.c:313 
(discriminator 2)

0x4202ba1f: lv_timer_handler at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/lvgl__lvgl/src/misc/lv_timer.c:109

0x42010184: lvgl_port_task at E:/esp_fryer_highend/ESP-fryer-highend/managed_components/espressif__esp_lvgl_port/esp_lvgl_port.c:441

0x4037e29d: vPortTaskWrapper at E:/esp_idf_5.0/Espressif/frameworks/esp-idf-v5.0/components/freertos/FreeRTOS-Kernel/portable/xtensa/port.c:151
```

**Comment by espzav** [comment]
Hi @hayschan, I tried today the latest version of the component `ws_7inch` and everything is ok. 
ws_7inch version: 1.0.1
LVGL port version: 1.1.0
GT911 version: 1.0.7~1
LCD touch version: 1.0.4
RA8875 version: 1.0.0

**Comment by hayschan** [comment]
I was wondering if you would be open to allowing me to give it a try as well? I understand that you've already tested it and I would be grateful for the opportunity to experiment with it on my own machine. 

Please let me know if it's okay with you and if there are any specific instructions or considerations I should keep in mind while trying out the code. I truly appreciate your time and willingness to share your work.

Thank you in advance! 😃

**Comment by hayschan** [comment]
> ws_7inch version: 1.0.1
> LVGL port version: 1.1.0
> GT911 version: 1.0.7~1
> LCD touch version: 1.0.4
> RA8875 version: 1.0.0

I just checked. My versions of the components are excatly the same as you.

**Comment by espzav** [comment]
HI @hayschan,
I am sorry for the late reply. Here is my testing project from SquareLine with latest `ws_7inch` component.
[test_800.zip](https://github.com/espressif/esp-bsp/files/11563983/test_800.zip)

**Comment by VojtechBartoska** [comment]
Any updates on this @hayschan? Thanks

**Comment by hayschan** [comment]
Sorry for the late reply. I will try the code tomorrow and update the findings.

**Comment by VojtechBartoska** [comment]
@hayschan Thanks a lot, looking forward for your feedback

**Comment by hayschan** [comment]
> I am sorry for the late reply. Here is my testing project from SquareLine with latest `ws_7inch` component.
[test_800.zip](https://github.com/espressif/esp-bsp/files/11563983/test_800.zip)

I have tried running `test_800.zip`. Runs flawlessly. 👍🏼 

I should now dig into my code. 

Thanks a lot for the help. ❤️

**Comment by VojtechBartoska** [comment]
@hayschan Great, closing this issue as solved.
